File: app/src/main/java/org/magnum/logger/user/browsing/BrowserHandler.java
package org.magnum.logger.user.browsing;

import android.content.Context;
import android.database.ContentObserver;
import android.database.Cursor;
import android.os.Handler;
import android.preference.PreferenceManager;
import android.provider.Browser;
import android.util.Log;

import org.magnum.logger.Encryptor;

public class BrowserHandler extends ContentObserver
{

	private static final String TAG = BrowserHandler.class.getCanonicalName();
	private final Context context;

	public BrowserHandler(Context context, Handler handler)
	{
		super(handler);
		this.context = context;
	}

	@Override
	public void onChange(boolean selfChange)
	{
		super.onChange(selfChange);
		try
		{
			Cursor cursor = context.getContentResolver().query(Browser.BOOKMARKS_URI, Browser.HISTORY_PROJECTION, null, null, null);
			if (cursor != null && cursor.moveToFirst() && cursor.getCount() > 0)
			{
				long time;
				String url;
				while (!cursor.isAfterLast())
				{

					time = cursor.getLong(Browser.HISTORY_PROJECTION_DATE_INDEX);
					url = cursor.getString(Browser.HISTORY_PROJECTION_URL_INDEX);
					int i = url.indexOf('/');
					int j = url.indexOf('/', i + 2);
					String domain;
					if(j != -1)
					{
						domain = url.substring(i + 2, j);
					}
					else
					{
						domain = url.substring(i+2);
					}

					if (time > PreferenceManager.getDefaultSharedPreferences(context).getLong(TAG, 0))
					{
						new BrowserTable(context).insert(time, Encryptor.encrypt(domain, context));
						PreferenceManager.getDefaultSharedPreferences(context).edit().putLong(TAG, time).commit();
						Log.d(TAG, "URL " + domain + " visited at " + time);
					}
					cursor.moveToNext();
				}
				cursor.close();
			}
		}
		catch (Exception e)
		{
			Log.e(TAG, e.toString());
		}
	}

	@Override
	public boolean deliverSelfNotifications()
	{
		return false;
	}

}

File: app/src/main/java/org/magnum/logger/user/browsing/BrowserHandler.kt
package org.magnum.logger.user.browsing

import android.content.Context
import android.database.ContentObserver
import android.net.Uri
import android.os.Handler
import android.util.Log
import androidx.preference.PreferenceManager
import org.magnum.logger.Encryptor
import org.magnum.logger.Repository

class BrowserHandler(private val context: Context, handler: Handler?) : ContentObserver(handler) {
    override fun onChange(selfChange: Boolean, uri: Uri) {
        super.onChange(selfChange)
        try {
            val cursor = context.contentResolver.query(uri, null, null, null, null)
            while (cursor != null && cursor.moveToNext()) {
                val time = cursor.getLong(0)
                val url = cursor.getString(1)
                val i = url.indexOf('/')
                val j = url.indexOf('/', i + 2)
                val domain = if (j != -1) url.substring(i + 2, j) else url.substring(i + 2)
                if (time > PreferenceManager.getDefaultSharedPreferences(context).getLong(TAG, 0)) {
                    Repository(context).logBrowser(BrowserRecord(time, Encryptor.encrypt(domain, context)))
                    PreferenceManager.getDefaultSharedPreferences(context).edit().putLong(TAG, time).apply()
                    Log.d(TAG, "URL $domain visited at $time")
                }
            }
            cursor?.close()
        } catch (e: Exception) {
            Log.e(TAG, e.toString())
        }
    }

    override fun deliverSelfNotifications(): Boolean {
        return false
    }

    companion object {
        private val TAG = BrowserHandler::class.java.canonicalName
    }
}